### PR TITLE
added API user details

### PR DESCRIPTION
**Pull request by vinyll**
I'm not happy about having a 302 when the wrong token is provided.
I assume that the middleware should raise a 4xx (403?).

**Opinions are welcome**

### CODE REVIEW

File: api/tests.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        )
+        response = self.client.get('/api/profile/',
+                                   HTTP_AUTHORIZATION='Bearer {}'.format(token)
+                                   )
+        self.assertEquals(response.status_code, 200)
+        self.assertEquals(response.json()['id'], self.test_user.id)
+        self.assertEquals(response.json()['email'], self.test_user.email)
+
+    def test_access_with_wrong_token(self):
+        AccessToken.objects.create(
```
Comment (davidbgk) [inline api/tests.py:28]: Is the token creation necessary in that case?

File: api/tests.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    def test_profile_data(self):
+        token = AccessToken.objects.create(
+            user=self.test_user, token='1234567890',
+            application=self.application,
+            expires=timezone.now() + datetime.timedelta(days=1),
+            scope='read'
+        )
+        response = self.client.get('/api/profile/',
+                                   HTTP_AUTHORIZATION='Bearer {}'.format(token)
+                                   )
```
Comment (davidbgk) [inline api/tests.py]: Not fond of that indentation, proposition:

```
response = self.client.get(
    '/api/profile/', HTTP_AUTHORIZATION='Bearer {}'.format(token))
```

File: api/tests.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    def test_access_with_wrong_token(self):
+        AccessToken.objects.create(
+            user=self.test_user, token='1234567890',
+            application=self.application,
+            expires=timezone.now() + datetime.timedelta(days=1),
+            scope='read'
+        )
+        response = self.client.get('/api/profile/',
+                                   HTTP_AUTHORIZATION='Bearer 123')
+        self.assertEquals(response.status_code, 302)
```
Comment (davidbgk) [inline api/tests.py]: To me it should return a `401` in that case.

File: api/views.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+from django.contrib.auth.decorators import login_required
+
+from oauth2_provider.views import ScopedProtectedResourceView
+
+from core.decorators import class_decorator
+
+
+@class_decorator(login_required)
+class ProfileView(ScopedProtectedResourceView):
+    required_scopes = ['read']
```
Comment (davidbgk) [inline api/views.py:11]: Can we have an explicit `email` scope when it returns the email?

**Review by davidbgk — CHANGES_REQUESTED** [review]
👍

File: api/tests.py (hunk 1 above)
Comment (vinyll) [reply api/tests.py:28]: It shouldn't, unless it would consider any token; and that's part of the test ;)

File: api/tests.py (hunk 3 above)
Comment (vinyll) [reply api/tests.py]: Or a 403
cf https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.2

> The response MUST include a WWW-Authenticate header field

File: api/views.py (hunk 4 above)
Comment (vinyll) [reply api/views.py:11]: Tha'ts planned in a separate PR :)

File: api/tests.py (hunk 3 above)
Comment (davidbgk) [reply api/tests.py]: I don't get the issue here, why not [adding it](http://stackoverflow.com/questions/8341763/proper-www-authenticate-header-for-oauth-provider/33902917#33902917)?

Comment (vinyll) [reply api/tests.py]: It may be the way to go. However it will be reconsidering the middleware that django-oauth-toolkit provides. Tho I'm ok with that ;)

Comment (davidbgk) [reply api/tests.py]: 🙄 http://timperrett.com/2016/11/12/frameworks-are-fundimentally-broken/

**Comment by vinyll** [comment]
It's now returning a 401 with `www-authenticate`.
The indentation was changed according to you suggestion.

File: oauth/middleware.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        self.get_response = get_response
+
+    def __call__(self, request):
+        if request.META.get('HTTP_AUTHORIZATION', '').startswith('Bearer'):
+            if not hasattr(request, 'user') or request.user.is_anonymous():
+                user = authenticate(request=request)
+                if user:
+                    request.user = request._cached_user = user
+                else:
+                    response = HttpResponse('Unauthorized', status=401)
```
Comment (davidbgk) [inline oauth/middleware.py:21]: That's “funny” that Django [doesn't have a custom HttpResponse](https://code.djangoproject.com/ticket/6349#comment:8) for that.

**Review by davidbgk — APPROVED** [review]

File: oauth/middleware.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
+from django.contrib.auth import authenticate
+from django.http import HttpResponse
+
+
+class OAuth2TokenMiddleware:
+    """
+    Django 1.10 has changed the middlewares signature.
+    oauth2_provider.middleware.OAuth2TokenMiddleware is still old style.
+    https://github.com/evonove/django-oauth-toolkit/issues/406
```
Comment (yohanboniface) [inline oauth/middleware.py:9]: Fix committed: https://github.com/evonove/django-oauth-toolkit/commit/911f7b828ec70ca02aab111a660412bd1e90a7c2 (but not yet released).

Comment (vinyll) [reply oauth/middleware.py:9]: Well, in fact I should even delete this comment as we inspired from but finally enriched this middleware

Comment (yohanboniface) [reply oauth/middleware.py:9]: Humm, this still looks like a hack to me anyway. It seems weird to me to be using a third party oauth lib, and needing to:
- check for "Bearer" by hand (not sure this work for all oauth flows)
- set request.user by hand
- return a 401 when not auth
- deal with WWW-Authenticate header

Imho, either this should be committed upstream, either we should change third party app, either we should review our needs.
What do you think?

**Pull request merged**